File: typescript/src/native.ts
import { spawn, type ChildProcess } from 'node:child_process';
import { L2S1Client, L2S1Error, positiveTimeout } from './http.js';
import type { CallOptions } from './http.js';
import type { Capabilities, DecisionPolicy, DecisionRequest, DecisionResponse } from './types.js';
import { resolveRuntime } from './runtime.js';

export interface LoadOptions {
  model: string;
  /** Override the platform package with a custom Rust executable or a name on PATH. */
  binaryPath?: string;
  device?: 'cpu' | 'cuda' | 'metal';
  mmproj?: string;
  lora?: string;
  context?: number;
  batch?: number;
  ubatch?: number;
  threads?: number;
  gpuLayers?: number;
  executionMode?: 'fresh' | 'prefix-reuse' | 'state-restore' | 'parallel';
  parallelWidth?: number;
  promptLayout?: 'legacy' | 'state-first';
  promptDetail?: 'minimal' | 'typed' | 'typed-examples';
  policy?: DecisionPolicy;
  /** Model startup timeout; defaults to 120 seconds. */
  startupTimeoutMs?: number;
  /** Per HTTP request timeout; defaults to 180 seconds. */
  timeoutMs?: number;
  signal?: AbortSignal;
  /** Additional Rust CLI flags. --listen is reserved for process ownership. */
  extraArgs?: string[];
  /** Native diagnostic chunks. Kept draining even when no callback is supplied. */
  onStderr?: (chunk: string) => void;
}

function argumentsFor(options: LoadOptions): string[] {
  if (!options.model?.trim()) throw new TypeError('model is required');
  const extra = options.extraArgs ?? [];
  if (extra.some((arg) => arg === '--listen' || arg.startsWith('--listen='))) {
    throw new TypeError('--listen is managed by L2S1.load');
  }
  const args = ['--model', options.model];
  const flags: [string, string | number | undefined][] = [
    ['device', options.device], ['mmproj', options.mmproj], ['lora', options.lora],
    ['context', options.context], ['batch', options.batch], ['ubatch', options.ubatch],
    ['threads', options.threads], ['gpu-layers', options.gpuLayers], ['execution-mode', options.executionMode],
    ['parallel-width', options.parallelWidth], ['prompt-layout', options.promptLayout], ['prompt-detail', options.promptDetail],
    ['min-top-probability', options.policy?.min_top_probability], ['min-candidate-mass', options.policy?.min_candidate_mass],
  ];
  for (const [key, value] of flags) if (value !== undefined) args.push(`--${key}`, String(value));
  // Port 0 lets Rust bind an available port atomically, without a port reservation race.
  args.push(...extra, '--listen', '127.0.0.1:0');
  return args;
}

/** Owns one resident Rust model process. Requests use the existing HTTP v1 API. */
export class RustProcessBackend implements AsyncDisposable {
  private closing: Promise<void> | undefined;
  private exited = false;
  private readonly exit: Promise<void>;
  private readonly lifetime = new AbortController();
  private client: L2S1Client | undefined;
  private constructor(private readonly child: ChildProcess) {
    this.exit = new Promise((resolve) => {
      child.once('close', () => {
        this.exited = true;
        this.lifetime.abort(new L2S1Error('Rust process closed', 'process_closed'));
        resolve();
      });
    });
  }
  static async load(options: LoadOptions): Promise<RustProcessBackend> {
    options.signal?.throwIfAborted();
    const args = argumentsFor(options);
    const startupTimeout = positiveTimeout(options.startupTimeoutMs ?? 120_000);
    positiveTimeout(options.timeoutMs ?? 180_000);
    const runtime = resolveRuntime(options.binaryPath, options.device);
    const child = spawn(runtime.binaryPath, args, { env: runtime.env, stdio: ['ignore', 'ignore', 'pipe'], windowsHide: true, shell: false });
    const engine = new RustProcessBackend(child);
    let diagnosticTail = '';
    let pending = '';
    let reportReady: (url: string) => void = () => {};
    let reportFailure: (error: unknown) => void = () => {};
    let announced = false;
    const ready = new Promise<string>((resolve, reject) => { reportReady = resolve; reportFailure = reject; });
    child.once('error', (cause) => reportFailure(new L2S1Error(`Could not start Rust executable: ${cause.message}`, 'spawn_failed', undefined, undefined, undefined, { cause })));
    child.once('close', (code, signal) => reportFailure(new L2S1Error(
      `Rust process exited during startup (${code ?? signal}): ${diagnosticTail}`, 'startup_failed')));
    child.stderr!.setEncoding('utf8');
    child.stderr!.on('data', (chunk: string) => {
      diagnosticTail = (diagnosticTail + chunk).slice(-8192);
      if (options.onStderr) {
        try { options.onStderr(chunk); } catch (cause) { reportFailure(cause); }
      }
      if (announced) return;
      pending += chunk;
      const lines = pending.split('\n');
      pending = lines.pop()!.slice(-8192);
      for (const line of lines) {
        const match = /^l2s1 HTTP listening on 127\.0\.0\.1:(\d+)\s*$/.exec(line);
        if (match && Number(match[1]) > 0 && Number(match[1]) <= 65535) {
          announced = true;
          reportReady(`http://127.0.0.1:${match[1]}`);
          break;
        }
      }
    });
    const startup = new AbortController();
    const timer = setTimeout(() => startup.abort(new L2S1Error('Rust model startup timed out', 'startup_timeout')), startupTimeout);
    const signal = options.signal ? AbortSignal.any([options.signal, startup.signal]) : startup.signal;
    const onAbort = () => reportFailure(signal.reason);
    signal.addEventListener('abort', onAbort, { once: true });
    try {
      const baseUrl = await ready;
      engine.client = new L2S1Client({ baseUrl, timeoutMs: options.timeoutMs ?? 180_000 });
      // A listening announcement follows bind; health confirms the API responds.
      await engine.client.health({ signal: AbortSignal.any([signal, engine.lifetime.signal]) });
      signal.throwIfAborted();
      return engine;
    } catch (error) {
      await engine.close();
      throw error;
    } finally {
      clearTimeout(timer);
      signal.removeEventListener('abort', onAbort);
    }
  }
  private callOptions(options: CallOptions): CallOptions {
    if (this.closing || this.exited || !this.client) throw new L2S1Error('Rust process is closed', 'process_closed');
    return { ...options, signal: options.signal ? AbortSignal.any([options.signal, this.lifetime.signal]) : this.lifetime.signal };
  }
  decide(request: DecisionRequest, options: CallOptions = {}): Promise<DecisionResponse> {
    return this.client!.decide(request, this.callOptions(options));
  }
  capabilities(options: CallOptions = {}): Promise<Capabilities> {
    return this.client!.capabilities(this.callOptions(options));
  }
  /** Idempotent. Aborts local HTTP calls, terminates the owned child and waits for exit. */
  close(): Promise<void> {
    if (!this.closing) {
      this.lifetime.abort(new L2S1Error('Rust process closed', 'process_closed'));
      this.closing = (async () => {
        if (this.exited) return;
        this.child.kill('SIGTERM');
        const timer = setTimeout(() => { this.child.kill('SIGKILL'); }, 2000);
        try { await this.exit; } finally { clearTimeout(timer); }
      })();
    }
    return this.closing;
  }
  [Symbol.asyncDispose](): Promise<void> { return this.close(); }
}

File: sdks/python/src/l2s1/native.py
"""Own one resident Rust process; reuse the TypeScript runtime bundle format."""
from __future__ import annotations

import asyncio
import hashlib
import json
import os
import platform
import re
import shutil
import sys
from collections.abc import Callable, Sequence
from dataclasses import dataclass, field
from pathlib import Path
from typing import Literal

from .backend import L2S1Error, positive_timeout
from .http import L2S1Client
from .stdio import StdioClient
from .models import Capabilities, DecisionPolicy, DecisionRequest, DecisionResponse

VERSION = "0.1.1"
Device = Literal["cpu", "cuda", "metal"]
ExecutionMode = Literal["fresh", "prefix-reuse", "state-restore", "parallel"]


@dataclass(frozen=True)
class LoadOptions:
    model: str
    transport: Literal["stdio", "http"] = "stdio"
    binary_path: str | None = None
    runtime_dir: str | Path | None = None
    device: Device = "cpu"
    mmproj: str | None = None
    lora: str | None = None
    context: int | None = None
    batch: int | None = None
    ubatch: int | None = None
    threads: int | None = None
    gpu_layers: int | None = None
    execution_mode: ExecutionMode | None = None
    parallel_width: int | None = None
    prompt_layout: Literal["legacy", "state-first"] | None = None
    prompt_detail: Literal["minimal", "typed", "typed-examples"] | None = None
    policy: DecisionPolicy | None = None
    startup_timeout_ms: int = 120_000
    timeout_ms: int = 180_000
    extra_args: tuple[str, ...] = field(default_factory=tuple)
    on_stderr: Callable[[str], None] | None = None


def _arguments(options: LoadOptions) -> list[str]:
    if not options.model.strip():
        raise ValueError("model is required")
    if options.device not in {"cpu", "cuda", "metal"}:
        raise ValueError("unsupported device")
    if any(arg == "--stdio" or arg.startswith("--stdio=") or arg == "--listen" or arg.startswith("--listen=") for arg in options.extra_args):
        raise ValueError("--stdio/--listen are managed by L2S1.load")
    if options.transport not in {"stdio", "http"}:
        raise ValueError("unsupported transport")
    positive_timeout(options.startup_timeout_ms)
    positive_timeout(options.timeout_ms)
    args = ["--model", options.model]
    if options.device != "cpu":
        args.extend(["--device", options.device])
    for name in ("mmproj", "lora", "context", "batch", "ubatch", "threads", "gpu_layers", "execution_mode",
                 "parallel_width", "prompt_layout", "prompt_detail"):
        value = getattr(options, name)
        if value is not None:
            args.extend(["--" + name.replace("_", "-"), str(value)])
    if options.policy is not None:
        policy = DecisionPolicy.model_validate(options.policy)
        args.extend(["--min-top-probability", str(policy.min_top_probability),
                     "--min-candidate-mass", str(policy.min_candidate_mass)])
    return [*args, *options.extra_args, *( ["--listen", "127.0.0.1:0"] if options.transport == "http" else ["--stdio"] )]


def _resolve_runtime(options: LoadOptions) -> tuple[str, dict[str, str]]:
    env = dict(os.environ)
    if options.binary_path is not None:
        if options.runtime_dir is not None:
            raise ValueError("choose binary_path or runtime_dir")
        if not options.binary_path.strip():
            raise ValueError("binary_path must not be empty")
        return options.binary_path, env
    if options.runtime_dir is None:
        executable = shutil.which("l2s1")
        if executable is None:
            raise L2S1Error("Supply binary_path or a TypeScript runtime bundle runtime_dir, or install l2s1 on PATH",
                            "runtime_not_installed")
        return executable, env
    root = Path(options.runtime_dir).resolve()
    manifest = json.loads((root / "runtime-manifest.json").read_text(encoding="utf-8"))
    machine = platform.machine().lower()
    arch = {"x86_64": "x64", "amd64": "x64", "aarch64": "arm64", "arm64": "arm64"}.get(machine, machine)
    host = {"linux": "linux", "darwin": "darwin", "win32": "win32"}.get(sys.platform, sys.platform)
    if manifest.get("platform") != f"{host}-{arch}":
        raise L2S1Error("Runtime bundle does not match this OS/architecture", "unsupported_platform")
    if manifest.get("version") != VERSION:
        raise L2S1Error(f"Runtime version mismatch: expected {VERSION}", "runtime_version_mismatch")
    if options.device not in manifest.get("devices", []):
        raise L2S1Error("Runtime bundle does not support the requested device", "unsupported_device")
    executable_name = "l2s1.exe" if host == "win32" else "l2s1"
    files = manifest.get("files", [])
    if not isinstance(files, list) or not any(item.get("path") == f"bin/{executable_name}" for item in files):
        raise L2S1Error("Runtime manifest is missing its executable", "invalid_runtime")
    for item in files:
        path = (root / item["path"]).resolve()
        if not path.is_relative_to(root / "bin") or not path.is_file():
            raise L2S1Error("Runtime manifest contains a missing or invalid path", "invalid_runtime")
        with path.open("rb") as stream:
            digest = hashlib.file_digest(stream, "sha256").hexdigest()
        if path.stat().st_size != item["bytes"] or digest != item["sha256"]:
            raise L2S1Error("Runtime bundle checksum mismatch", "invalid_runtime")
    variable = "PATH" if host == "win32" else "DYLD_LIBRARY_PATH" if host == "darwin" else "LD_LIBRARY_PATH"
    if host == "win32":
        variable = next((key for key in env if key.lower() == "path"), variable)
    env[variable] = os.pathsep.join(value for value in (str(root / "bin"), env.get(variable, "")) if value)
    return str(root / "bin" / executable_name), env


class RustProcessBackend:
    def __init__(self, child: asyncio.subprocess.Process) -> None:
        self._child = child
        self._client: L2S1Client | StdioClient | None = None
        self._drain: asyncio.Task[None] | None = None
        self._closing: asyncio.Task[None] | None = None

    @classmethod
    async def load(cls, options: LoadOptions) -> RustProcessBackend:
        args = _arguments(options)
        executable, env = _resolve_runtime(options)
        try:
            child = await asyncio.create_subprocess_exec(executable, *args, env=env,
                                                        stdin=asyncio.subprocess.PIPE if options.transport == "stdio" else asyncio.subprocess.DEVNULL,
                                                        stdout=asyncio.subprocess.PIPE if options.transport == "stdio" else asyncio.subprocess.DEVNULL,
                                                        stderr=asyncio.subprocess.PIPE,
                                                        limit=64 * 1024 * 1024)
        except OSError as error:
            raise L2S1Error(f"Could not start Rust executable: {error}", "spawn_failed") from error
        backend = cls(child)
        ready: asyncio.Future[str] = asyncio.get_running_loop().create_future()

        async def drain() -> None:
            assert child.stderr is not None
            tail = ""
            pending = ""
            try:
                while chunk := await child.stderr.read(4096):
                    text = chunk.decode("utf-8", errors="replace")
                    tail = (tail + text)[-8192:]
                    if options.on_stderr is not None:
                        options.on_stderr(text)
                    if ready.done():
                        continue
                    pending += text
                    lines = pending.split("\n")
                    pending = lines.pop()[-8192:]
                    for line in lines:
                        if options.transport == "stdio" and line.strip() == "l2s1 stdio ready" and not ready.done():
                            ready.set_result("stdio")
                        match = re.fullmatch(r"l2s1 HTTP listening on 127\.0\.0\.1:(\d+)\s*", line)
                        if match and 0 < int(match[1]) <= 65535 and not ready.done():
                            ready.set_result(f"http://127.0.0.1:{match[1]}")
                if not ready.done():
                    ready.set_exception(L2S1Error(f"Rust process exited during startup: {tail}", "startup_failed"))
            except Exception as error:
                if not ready.done():
                    ready.set_exception(error)
                # A diagnostic callback must never stop draining the child pipe.
                while await child.stderr.read(4096):
                    pass

        backend._drain = asyncio.create_task(drain())
        try:
            async with asyncio.timeout(options.startup_timeout_ms / 1000):
                url = await ready
                backend._client = StdioClient(child, options.timeout_ms) if options.transport == "stdio" else L2S1Client(url, timeout_ms=options.timeout_ms)
                await backend._client.health()
            return backend
        except BaseException:
            await backend.close()
            raise

    def _open_client(self) -> L2S1Client | StdioClient:
        if self._closing is not None or self._child.returncode is not None or self._client is None:
            raise L2S1Error("Rust process is closed", "process_closed")
        return self._client

    async def decide(self, request: DecisionRequest, *, timeout_ms: int | None = None) -> DecisionResponse:
        return await self._open_client().decide(request, timeout_ms=timeout_ms)

    async def decide_batch(self, requests: Sequence[DecisionRequest], *, timeout_ms: int | None = None) -> list[DecisionResponse]:
        return await self._open_client().decide_batch(requests, timeout_ms=timeout_ms)

    async def capabilities(self, *, timeout_ms: int | None = None) -> Capabilities:
        return await self._open_client().capabilities(timeout_ms=timeout_ms)

    async def _close(self) -> None:
        if self._client is not None:
            await self._client.close()
        if self._child.returncode is None:
            try:
                self._child.terminate()
            except ProcessLookupError:
                pass
            try:
                async with asyncio.timeout(2):
                    await self._child.wait()
            except TimeoutError:
                try:
                    self._child.kill()
                except ProcessLookupError:
                    pass
                await self._child.wait()
        if self._drain is not None:
            await self._drain

    async def close(self) -> None:
        if self._closing is None:
            self._closing = asyncio.create_task(self._close())
        await asyncio.shield(self._closing)
